datalad/metadata/metadata.py

```diff
@@ -430,47 +967,94 @@ def __call__(
+            if define_key or (apply2global and
+                              (init or purge or reset or add or remove)):
+                yield get_status_dict(
+                    ds=ds,
+                    status='ok',
+                    metadata=db,
+                    **res_kwargs)
+            #
+            # file metadata manipulation
+            #
             ds_paths = [p['path'] for p in content]
-            if not dataset_global:
-                if reset or purge or add or init or remove:
-                    # file metadata manipulation
-                    mod_paths = []
-                    for mp in ds.repo.set_metadata(
-                            ds_paths,
-                            reset=reset,
-                            add=add,
-                            init=init,
-                            remove=remove,
-                            purge=purge,
-                            # we always go recursive
-                            # TODO is that a good thing? But how to otherwise distinuish
-                            # this kind of recursive from the one across datasets in
-                            # the API?
-                            recursive=True):
-                        if mp.get('success', False):
-                            mod_paths.append(mp['file'])
-                        else:
-                            yield get_status_dict(
-                                status='error',
-                                message='setting metadata failed',
-                                path=opj(ds.path, mp[0]),
-                                type='file',
-                                **res_kwargs)
-                    # query the actually modified paths only
-                    ds_paths = mod_paths
-
-                # and lastly, query -- even if we set before -- there could
-                # be side-effect from multiple set paths on an individual
-                # path, hence we need to query to get the final result
-                for file, meta in ds.repo.get_metadata(ds_paths):
-                    r = get_status_dict(
-                        status='ok',
-                        path=opj(ds.path, file),
-                        type='file',
-                        metadata=meta,
-                        **res_kwargs)
+            if not apply2global and (reset or purge or add or init or remove):
+                respath_by_status = {}
+                for res in ds.repo.set_metadata(
+                        ds_paths,
+                        reset=reset,
+                        add=add,
+                        init=init,
+                        remove=remove,
+                        purge=purge,
+                        # we always go recursive
+                        # XXX is that a good thing? But how to otherwise distinuish
+                        # this kind of recursive from the one across datasets in
+                        # the API?
```

Comment:
May be it could be distinguished by renaming the parameter for `AnnexRepo.set_metadata` to `traverse`?

Reply:
It is not about the name, it is about the inability to distinguish recursion from traversal when the respective boundaries are invisible.

Reply:
Good point. I missed that.